File: project/android/apps/MnnLlmApp/app/src/main/java/com/alibaba/mls/api/source/ModelSourceConfig.java
// Created by ruoyi.sjd on 2025/2/11.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.

package com.alibaba.mls.api.source;

import java.util.LinkedHashMap;

public class ModelSourceConfig {

    public LinkedHashMap<String, RepoConfig> repos;

    //we might create a server to store the model source configs, but now use some hack
    // to download both modelscope and huggingface models
    public static ModelSourceConfig createMockSourceConfig() {
        ModelSourceConfig config = new ModelSourceConfig();
        config.repos = new LinkedHashMap<>();
        return config;
    }

    public RepoConfig getRepoConfig(String modelId) {
        if (repos == null) {
            return null;
        }
        RepoConfig repoConfig = repos.get(modelId);
        if (repoConfig == null) {
            String[] splits = modelId.split("/");
            if (splits.length == 2) {
                String modelScopeId = "MNN/" + splits[1];
                repoConfig = new RepoConfig(modelScopeId, modelId, modelId);
            } else {
                String modelScopeId = "MNN/" + modelId;
                String huggingFaceId = "taobao-mnn/" + modelId;
                repoConfig = new RepoConfig(modelScopeId, huggingFaceId, huggingFaceId);
            }
            repos.put(modelId, repoConfig);
        }
        return repoConfig;
    }
}

File: apps/frameworks/model_downloader/android/src/main/java/com/alibaba/mls/api/source/ModelSourceConfig.kt
// Created by ruoyi.sjd on 2025/2/11.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.
package com.alibaba.mls.api.source

class ModelSourceConfig {
    var repos: LinkedHashMap<String, RepoConfig>? = null

    fun getRepoConfig(modelId: String): RepoConfig? {
        if (repos == null) {
            return null
        }
        var repoConfig = repos!![modelId]
        if (repoConfig == null) {
            val splits = modelId.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (splits.size == 2) {
                val modelScopeId = "MNN/" + splits[1]
                repoConfig = RepoConfig(modelScopeId, modelId, modelId)
            } else {
                val modelScopeId = "MNN/$modelId"
                val huggingFaceId = "taobao-mnn/$modelId"
                repoConfig = RepoConfig(modelScopeId, huggingFaceId, huggingFaceId)
            }
            repos!![modelId] = repoConfig
        }
        return repoConfig
    }

    companion object {
        //we might create a server to store the model source configs, but now use some hack
        // to download both modelscope and huggingface models
        fun createMockSourceConfig(): ModelSourceConfig {
            val config = ModelSourceConfig()
            config.repos = LinkedHashMap()
            return config
        }
    }
}
